README.md
# Puzzle Tokens
Puzzle Tokens is a Sketch plugin to specify and apply design tokens (in LESS or SCSS format) to Sketch layers (with shared styles) with auto styles preview generation.

### [Check this article](https://medium.com/@akalenyuk/bringing-styles-preprocessing-to-sketch-9cdf0d0c49bd) to get a detailed overview with examples and join [Discussions](https://github.com/ingrammicro/puzzle-tokens/discussions) for live talk

## Installation
1. Download [zip file](https://github.com/ingrammicro/puzzle-tokens/raw/master/PuzzleTokens.sketchplugin.zip)
2. Unarchive and install
3. Download and install [Node.js](https://nodejs.org/en/download/)
4. Instal _less_ or _sass_ using the following Terminal commands:
```
sudo -s
npm i less -g
npm i sass -g
```
5. Now you're good to go!

## Usage
1. Download [Example 1](https://github.com/ingrammicro/puzzle-tokens/tree/master/Styles/Tests/Buttons).
2. Open Library.sketch file in Sketch.app
3. Run Plugins > Puzzle Tokens > Apply Design Tokens menu command
4. Specify LESS file according to screenshot

<img width="755" src="https://raw.githubusercontent.com/ingrammicro/puzzle-tokens/master/Styles/Tests/Buttons/Illustration.png"/>

5. Repeat the same operation, but select "tokens-blue.less" file. See how styles and widgets look now.


## Features
The following CSS styles are supporting.
```css
// Text Layers

.TextStyle {
    // Sketch only styles
    pt-paragraph-spacing:  23.33;
    pt-text:               "New text content";
    -pt-text-size-behaviour: fixed-size;    // or "auto-height" or  "auto-width"           

    // CSS native styles
    font-size:             12px;   
    font-family:           "Open Sans";     // or "Open Sans","Times New Roman"
    font-weight:           bold;            // or extra-light, light, regular, medium, semibold, bold OR fa-solid (special trick to define FontAwesome solid icon)
    font-style:            italic;          // or normal
    line-height:           1.0;             // or 1.2 or 1.5 or any other mulitplier for font-size OR 10px
    color:                 #FFFFFF;         // HEX value OR any other CSS-compatible color values, such as red or black OR color variable name in double quotes
    opacity:               63%;             // supported "63%" or "0.42"
    text-transform:        uppercase;       // "uppercase", "lowercase", "none"
    text-decoration:       underline;       // "underline", "line-through"
    text-align:            left;            // "left", center", "right", "justify"
    vertical-align:        top;             // "top", "middle", "bottom"
    letter-spacing:        10px;            // <value>px OR "normal"
}

// Shape layers

.ShapeStyle {
    // SKETCH only properties
    border-position:       center;    // center OR inside OR outside
    border-line-end:            butt;      // butt / round / projecting
    border-line-join:           miter;     // miter / round / bevel
    border-start-arrowhead:     openarrow; // none / openarrow / filledarrow / opencircle / filledcircle / opensquare / filledsquare
    border-end-arrowhead:       openarrow; // none / openarrow / filledarrow / opencircle / filledcircle / opensquare / filledsquare            
    -pt-border-update:          true;  // Use it if you want to replace an existing border, instead of adding a new one.
    -pt-fill-update:            true;  // Use it if you want to replace an existing fill, instead of adding a new one.

    /// CSS native properties
    background-color:      #B0AFB1;
    background-color:      linear-gradient(45deg, #000000,#B0AFB4);
    background-color:      linear-gradient(134deg, #004B3A 0%, #2D8B61 51%, #9BD77E 100%);
    background-color:      "mycolors/color01"; // assign color variable
    opacity:               63%;       // "63%" or "0.42"
    border-color:          #000000;
    border-color:          linear-gradient(45deg, #000000,#B0AFB4);
    border-color:          linear-gradient(134deg, #004B3A 0%, #2D8B61 51%, #9BD77E 100%);
    border-color:          "mycolors/color01"; // assign color variable
    border-width:          2px;
    border-style:          dotted;    // dotted OR dashed

    box-shadow:            none; // Clear any existing shadows from the layer style.
    box-shadow:            0 10px 20px 2 #FF00FF;
    box-shadow:            inset 0 10px 20px 2 rgba(0,0,0,0.1);
    box-shadow:            0 10px 20px 2 rgba(0,0,0,0.1), inset 0 10px 20px 2 rgba(0,0,0,0.1);
    -pt-shadow-update:          true;  // Use it if you want to replace any previous shadows, instead of adding a new one.

    mix-blend-mode:         normal; // normal / darken / multiply / color-burn / lighten / screen / color-dodge / overlay / difference 
                                    // exclusion / hue / saturation / color / luminosity
     
    // !!ATTENTION!!
    // Shared styles don't include radius property,
    // still you can set the radius-border for a style.
    // Border radius will be reapplied to layers
    // through style assigned to it. You can also
    // apply it to the layers or symbols directly.

    border-radius:         5px;
    border-radius:         5px 5px 0 0;
}

// Text & Shape Layer Common Properties
.Style{
    // Sketch only styles
    -pt-layer-type:          text; // "layer" or "text" // Usefull to help PT to understnand the type of the layer
    -pt-pin-left:            true; // "true" or "false
    -pt-pin-right:           true; // "true" or "false          
    -pt-pin-top:             true; // "true" or "false
    -pt-pin-bottom:          true; // "true" or "false
    -pt-fix-size-height:     true; // "true" or "false
    -pt-fix-size-width:      true; // "true" or "false
    -pt-skip-missed:         true: // "true" or "false // Usefull to skip style if no layer
                                                       // found without an error
    -pt-resize-symbol:      true:  //  resize layer owner (symbol master) to a layer size

    // The following properties are not a part of shared styles.
    // Bu you can set these properties for a shared style.
    // These properties will be reapplied to layers
    // through style assigned to it. You can also
    // apply it to the layers or symbols directly.

    width:                  100px;
    height:                 100px;

    margin-top:             10px;
    margin-left:            20px;
    margin-right:           20px;
    margin-bottom:          10px;
    
    // By default, margin is set relative to the layer's containing artboard or page.
    // (Specifically, the layer is moved so that it is positioned inside the artboard/
    // page at the specified margin. In the example above, the top-left corner of the
    // layer is set to x:10 and y:20 of the containing artboard.)
    //
    // If you prefer to set the margin  relative to another layer (for example, to set 
    // the text layer of a button within the boundaries of the button's background
    // layer style), use the "-pt-margin-relative-to" property shown below.
    
    -pt-margin-relative-to: "layer name"; // set margin values relative to the specified
                                          // layer. The specified layer must be a sibling
                                          // (at the same level in a group, artboard, or
                                          // symbol) of the layer to which you are
                                          // applying the margin styles. If not specified,
                                          // margin will be set relative to the artboard
                                          // or page
    -pt-margin-resize:      true;         // if true, resize the "margin-relative-to"
                                          // layer to "fit" the size of the current layer,
                                          // plus the specified margin. This will make
                                          // the "margin-relative-to" layer surround
                                          // the current layer at exactly the requested
                                          // margin.


    -pt-valign:                     middle; // middle / top / bottom : align vertically inside a parent
}

// Group & SymbolMaster Properties
#Group{
    -pt-smartlayout:         LeftToRight; // LeftToRight OR HorizontallyCenter OR RightToLeft OR TopToBottom 
                                          // OR VerticallyCenter OR BottomToTop OR None
}

// Group, SymbolMaster and Artboard Properties
#Symbol{
    -pt-fit-content:        true;        // Resize to fit content
}

// SymbolMaster properties
#Symbol{
    -pt-resize-instances:   true;        // Resize all instances of a symbol; the same as
                                         // clicking Sketch's "Shrink instance to fit
                                         // content" button in the Overrides section of
                                         // the instance. (This reapplies SmartLayout,
                                         // useful when you change the size of a symbol.)
}

// Symbol Instance Properties
#Card .IconButton {                      // Path to a symbol-instance layer

    // You can change the symbol source for a symbol override (for example, to switch
    // the icon displayed in a symbol instance). Usage is:
    // -pt-override-symbol: ('AffectedLayerName', '#Path #To #New #Symbol');
    //
    // In the example shown here, the Card symbol has a symbol instance in a layer named
    // IconButton. If that IconButton instance has a symbol override in layer RightIcon,
    // you can set the icon to a new icon by sharing its Puzzle Tokens path like so:
    
    -pt-override-symbol:    ('RightIcon ', '#Icons #Arrows #RightArrow');
    
    // Set the path to 'none' to disable the symbol override and show nothing at all.
}

#Image{
    // Required Properties
    image:                 ~"images/new-logo.jpg";  // OR transparent
    // Optional Properties
    border-color:          white;
    border-width:          3px;
    box-shadow:            0 10px 20px 2 #FF00FF;
    width:                 100px;  // OR 50 %
    height:                100px;  // OR 50 %
    top:                   11px;
    bottom:                22px;
    left:                  2px;
    right:                 33px;
}
```


You can [look into more examples](https://github.com/ingrammicro/puzzle-tokens/tree/master/Styles/Tests) to get familiar with the plugin.

## Supported Targets
You can update the following Sketch objects.

```css
// Update shared style
.MyStyles .Group .Style1{
    color: #FFFFFF;
}

// Update symbol child style properties
#MySymbols #Buttons #Submit{
    .Text{
        color: #FFFFFF;
    }
}

// Update artboard child style properties
#MyArtboard #Group1{
    .Rectangle{
        color: #FFFFFF;
    }
}

// Assign shared style to artboard/symbol child
#MySymbols #Buttons #Submit{
    .Text{
        sktext-style: "MyStyle/Group/Style1";
    }
    .Back{
        sklayer-style: "MyStyle/Group/Back";
    }
}

// Update color variables  !!!!!!!! DONT USE GRADIENTS FOR COLOR VARIABLES. SKETCH WILL CRASH !!!!! 
.--COLORS-{
    .primary{
        color:                                      black;
    }
    .secondary{
        color:                                      grey;
    }
}

```
## Special Tricks for Selector Names
- To start a style or symbol name from a number just add --PT- prefix. So, "--PT-1 Style" will be transformed to "1 Style".
- To start a style or symbol name from a dot just use -DOT- special keyword. So, "-DOT-Style" will be transformed to ".Style"
- To use a space inside a style or symbol name you can uses a space (but formally it will not be a CSS compartible notation) or replace space by __ .So "My__First__Style" wil be transformed to "My First Style

## Required Style Properties
To apply _text_ style you need to defined at least one of the following properites:
- color
- font-family
- font-size
- font-weight
- text-transform
- text-align
- vertical-align
- pt-paragraph-spacing

For _layer_ style:
- background-color
- border-color
- box-shadow
- border-radius
- opacity

To create _image:
- image

To assign shared style to layer:
- sklayer-style
- sktext-style

## Sketch versions supported
Some styles, such as text font/transformation/etc require Sketch 53 or later. You're safe using the latest stable Sketch version to get the plugin working.


## Demos 
https://www.dropbox.com/s/f25cfsm26vgeu62/Recording1.mov?dl=0

## Command line API
```
#!/bin/bash
context="{\"file\":\"${HOME}/GitHub/puzzle-tokens/Styles/material-palettes/palettes.sketch\""
# The following line is optional
context="${context},\"saveAs\":\"${HOME}/GitHub/puzzle-tokens/Styles/material-palettes/palettes_new.sketch\""
context="${context},\"styles\":\"${HOME}/GitHub/puzzle-tokens/Styles/material-palettes/scss/palettes.scss\""
context="${context},\"commands\":\"apply,save,close\"}"
echo $context
/Applications/Sketch.app/Contents/Resources/sketchtool/bin/sketchtool --without-activating=YES --new-instance=No run ~/Library/Application\ Support/com.bohemiancoding.sketch3/Plugins/PuzzleTokens.sketchplugin "cmdRun" --context=$context
```

Compact variant:
```
/Applications/Sketch.app/Contents/Resources/sketchtool/bin/sketchtool --without-activating=YES --new-instance=No run ~/Library/Application\ Support/com.bohemiancoding.sketch3/Plugins/PuzzleTokens.sketchplugin "cmdRun" --context="{\"file\":\"${HOME}/GitHub/puzzle-tokens/Styles/material-palettes/palettes.sketch\",\"styles\":\"${HOME}/GitHub/puzzle-tokens/Styles/material-palettes/scss/palettes.scss\",\"commands\":\"apply,save,close\"}"
```


CHANGELOG.md
# Change Log
See discussions on https://github.com/ingrammicro/puzzle-tokens/discussions site

# Version 8.20.0 (21 Sep 2022)
Support for lessc custom path

# Version 8.19.7 (30 Aug 2022)
Bugfix

# Version 8.19.6 (23 Aug 2022)
Quck apply updates shadow too

# Version 8.19.5 (13 Jul 2022)
Corrected smartlayout option names

# Version 8.19.3 (15 Jun 2022)
Removed debug info

# Version 8.19.1 (16 May 2022)
Cmd-line fixed

# Version 8.19.0 (25 Apr 2022)
Split apply command to two commands

# Version 8.18.3 (24 Apr 2022)
Bug fixed

# Version 8.18.2 (13 Apr 2022)
Handle multi backgrounds correctly

# Version 8.18.0 (12 Apr 2022)
New settings added:
- Don't create new styles and ignore missed symbols
- Don't apply sizes and margins

# Version 8.17.5 (08 Nov 2021)
Minor issues fixed

# Version 8.17.4 (28 Oct 2021)
Fixed "@@blabla--token" correctly

# Version 8.17.1 (28 Sep 2021)
Process the following style value in better way
margin-top:                 @notification-padding + @tag-height-small + @notification-label-margin-top;

# Version 8.17.0 (12 Jul 2021)
Moved gradient parsing from own code to 3d-party library

# Version 8.16.4 (11 May 2021)
Updated to be compatible with Versions (by @vkoopmans-resoluut)

# Version 8.16.3 (18 Mar 2021)
Don't allow to use gradients in color variable value

# Version 8.16.2 (02 Mar 2021)
Fixed minor issues

# Version 8.16.1 (29 Jan 2021)
Fixed @import "@{brand}.less" 

# Version 8.16.0 (11 Jan 2021)
"border-position" activates border style
"-pt-valign: middle/bottom/top;" rule has been added
Other small improvements

# Version 8.15.0 (25 Nov 2020)
Now you can assign color variable to any style or layer color using double quotes to specify a color variable name.
Example:background-color:  "Red-Colors/red01";

Changed:
Now exporter name destination file as "puzzle-tokens.less" or "puzzle-tokens.scss" (#26 fixed)

Fixed issues:
- "red","green",etc color value can not be assigned to color var

# Version 8.14.4 (29 Oct 2020)
Supported color var groups

# Version 8.14.2 (29 Oct 2020)
Fixed opacity for text layers

# Version 8.14.1 (29 Oct 2020)
Fixed color var updating
Fixed support for image path in SASS/SCSS

# Version 8.14.0 (28 Oct 2020)
Added support for color variables
Example: https://github.com/ingrammicro/puzzle-tokens/tree/master/Styles/Releases/8.14.0

# Version 8.13.0 (26 Oct 2020)
Added support for mix-blend-mode CSS rule. 
Example: https://github.com/ingrammicro/puzzle-tokens/tree/master/Styles/Releases/8.13.0

##  Version 8.12.1 (10 Oct 2020)
Author: Josh Clark  (https://github.com/bigmedium)
Height/width dimensions have to be adjusted first, or bottom-margin and right-margin are incorrect.

##  Version 8.12.0 (15 Aug 2020)  (Цой жив)
Added configuration setting to define a custom path to SASS module

##  Version 8.11.0 (11 Aug 2020) 
Author: Josh Clark  (https://github.com/bigmedium)
1) Added support for gradients in border-color
2) When border-style is "none", apply no border style (or remove any existing border style if -pt-border-update is true)

##  Version 8.10.0 (6 Aug 2020) 
Author: Josh Clark  (https://github.com/bigmedium)
This update adds four new properties, which are specifically intended to be helpful for managing the spacing and size within buttons, for example.

Margin enhancements:

-pt-margin-relative-to: "layer name";
This sets margin values relative to the specified layer. The specified layer must be a sibling at the same level in a group, artboard, or symbol) of the layer to which you are applying the margin styles. If not specified, margin will be set relative to the artboard or page (which is the current behavior).

-pt-margin-resize: true;
If true, resize the "margin-relative-to" layer to "fit" the size of the current layer, plus the specified margin. This will make the "margin-relative-to" layer surround the current layer at exactly the requested margin. (Very helpful for effectively setting the padding of buttons and sizing the surrounding background layer to fit.)

Artboard/symbol resizing:

-pt-fit-content: true;
Applies only to artboards and symbol masters: if true, resize the artboard/symbol to fit its content.

-pt-resize-instances: true;
Applies only to symbol masters. If true, resize all instances of a symbol; the same as clicking Sketch's "Shrink instance to fit content" button in the Overrides section of the instance. (This reapplies SmartLayout, useful when you change the size of a symbol.)

##  Version 8.9.0 (26 June 2020)
Added new style 
-pt-text-size-behaviour: fixed-size; // or auto-height or auto-width

##  Version 8.8.4 (6 June 2020)
Hotfix

##  Version 8.8.3 (5 June 2020)
Improved diagnostic for missed Node.ls
User can change path to Node.js in Settings

##  Version 8.8.2 (4 June 2020)
Added ability to enable debug logging

##  Version 8.8.1 (2 June 2020)
Improved async mode for sending statistics

##  Version 8.8.0 (1 June 2020)
New ability:
`-pt-smartlayout: none` clears any smart layout on a symbol master or group (by @bigmedium)
Otjer:
Improved integration with Puzzle Publisher (many fixes in LESS tokens inspection)


##  Version 8.7.4 (28 May 2020)
Handle @import manually to inject tokens into all imported files

##  Version 8.7.3 (26 May 2020)
Update file protocol between Puzzle tokens and Puzzle Publisher

##  Version 8.7.2 (18 May 2020)
New text style should not get "default" borders (reported by @ed-mcry)

##  Version 8.7.1 (17 May 2020)
Disabled debug (hotfix for 8.7.0)

##  Version 8.7.0 (16 May 2020)
Changes:
1) If "box-shadow:none", clear any existing shadows from the layer style (by @bigmedium)
2) Added "-pt-shadow-update: true" property. Use it if you want to replace any previous shadows, instead of adding a new one. (by @bigmedium)
3) PT now sends anonymous usage data (using Google Analytics). You can disable it in Settings. But we need it enabled to plan PT future. Thanks.

##  Version 8.6.0 (05 May 2020)
Fixed "image:" property
Added support for SVG images

PuzzleTokens.sketchplugin/Contents/Sketch/constants.js
var Constants = {
    DOCUMENT_VERSION: "docVersion",
    TAB_SIZE: 2,
    HOTSPOT_PADDING: 0,
    LOGGING: false,
    RESOURCES_FOLDER: "scripts",
    ASSETS_FOLDER_PREFIX: "_pt-assets",
    SYMBOLTOKENFILE_POSTFIX: "inspector.json",
    CSSFILE_POSTFIX: "viewer.css",
    VARSFILE_POSTFIX: "vars.json",
    SASSFILE_POSTFIX: "vars.scss",
    SYMBOLPAGE_NAME: "Symbols",
    EXPORT_FORMAT_LESS: 0,
    EXPORT_FORMAT_SCSS: 1,

    SITE_CHANGELOG_URL: "https://github.com/ingrammicro/puzzle-tokens/blob/master/CHANGELOG.md",
    NODEJS_PATH: "/usr/local/bin/node",
    DEF_SASSMODULE_PATH: "/usr/local/lib/node_modules",
    DEF_LESSC_PATH:"/usr/local/bin/lessc",

    GA_ID: "UA-84277242-4"
};

var SettingKeys = {
    PLUGIN_PREVIEWER_DEF: "pluginPreviewerDef",
    PLUGIN_LAST_ONLY_UPDATE: "PLUGIN_LAST_ONLY_UPDATE",
    PLUGIN_PATH_TO_TOKENS_LESS: "pluginPathToTokensLess",
    PLUGIN_PATH_TO_TOKENS_LESS_LIST: "pluginPathToTokensLessList",
    PLUGIN_PATH_TO_TOKENS: "PLUGIN_PATH_TO_TOKENS",
    PLUGIN_PATH_TO_TOKENS_LIST: "PLUGIN_PATH_TO_TOKENS_LIST",
    PLUGIN_GENERATE_SYMBOLTOKENS: "pluginGenSymbTokens",
    PLUGIN_APPLY_CLEAR: "pluginApplyClear",
    PLUGIN_SHOW_JSON: "pluginDebug",
    PLUGIN_SHOW_DOUBLESTYLES: "pluginDoubleStyles",
    PLUGIN_APPLY_IGNORE_MISSED: "PLUGIN_APPLY_IGNORE_MISSED",
    PLUGIN_APPLY_SKIP_SIZES: "PLUGIN_APPLY_SKIP_SIZES",
    PLUGIN_APPLY_ONLY_UPDATE_STYLES: "PLUGIN_APPLY_ONLY_UPDATE_STYLES",
    PLUGIN_CREATE_SYMBOLS: "pluginCreateSymbols",
    PLUGIN_EXPORT_PATH_TO: "pluginExportPathTo",
    PLUGIN_EXPORT_OPEN_FINDER: "PLUGIN_EXPORT_OPEN_FINDER",
    PLUGIN_EXPORT_FORMAT: "pluginExportFormat",
    PLUGIN_EXPORT_OPTS: "pluginExportOpts",
    PLUGIN_EXPORT_LIB_STYLES: "pluginExportLibStyles",

    PLUGIN_NODEJS_PATH: "pluginNodeJSPath0",
    PLUGIN_SASSMODULE_PATH: "pluginSassModulePath",
    PLUGIN_LESSC_PATH: "PLUGIN_LESSC_PATH",

    PLUGIN_INSTALLED: "pluginInstalled",
    PLUGIN_GA_DISABLED: "pluginGADisabled",
    PLUGIN_LOGDEBUG_ENABLED: "pluginLogDebugEnabled",
};

const TRACK_INSTALLED = "installed"
const TRACK_STARTED = "started"
const TRACK_APPLY_DIALOG_SHOWN = "apply-dialog-shown"
const TRACK_APPLY_DIALOG_CLOSED = "apply-dialog-closed"  // cmd:ok,cancel
const TRACK_APPLY_COMPLETED = "apply-completed" // num_cs: created styles, num_us: updated styles, num_l: updated layers, quick: yes/no
const TRACK_APPLY_COMPLETED_LESS = "apply-completed-less"
const TRACK_APPLY_COMPLETED_SCSS = "apply-completed-scss"
const TRACK_EXPORT_COMPLETED = "export-completed"

var Sketch = require('sketch/dom')
var Settings = require('sketch/settings')
var Style = require('sketch/dom').Style
var Image = require('sketch/dom').Image
const Group = require('sketch/dom').Group
var SharedStyle = require('sketch/dom').SharedStyle
var UI = require('sketch/ui')
const path = require('path');
const Text = require('sketch/dom').Text
const Shape = require('sketch/dom').Shape
const Page = require('sketch/dom').Page
const SmartLayout = require('sketch').SmartLayout
const Artboard = require('sketch/dom').Artboard
const Document = require('sketch/dom').Document
const SymbolMaster = require('sketch/dom').SymbolMaster
const Rectangle = require('sketch/dom').Rectangle


const SKLAYER_STYLE = "sklayer-style"
const SKTEXT_STYLE = "sktext-style"
const PT_TEXT = "pt-text"
const PT_RESIZE_SYMBOL = "-pt-resize-symbol"
const PT_PARAGRAPH_SPACING = "pt-paragraph-spacing"
const PT_LAYER_TYPE = "-pt-layer-type"
const PT_SMARTLAYOUT = "-pt-smartlayout"
const PT_FIX_SIZE_HEIGHT = "-pt-fix-size-height"
const PT_FIX_SIZE_WIDTH = "-pt-fix-size-width"
const PT_SIZE_BEHAVIOUR = "-pt-text-size-behaviour"
const PT_PIN_LEFT = "-pt-pin-left"
const PT_PIN_RIGHT = "-pt-pin-right"
const PT_PIN_TOP = "-pt-pin-top"
const PT_PIN_BOTTOM = "-pt-pin-bottom"
const PT_BORDER_UPDATE = "-pt-border-update"
const PT_FILL_UPDATE = "-pt-fill-update"
const PT_SKIP_MISSED = "-pt-skip-missed"
const PT_SHADOW_UPDATE = "-pt-shadow-update"
const PT_MARGIN_RELATIVE_TO = "-pt-margin-relative-to"
const PT_MARGIN_RESIZE = "-pt-margin-resize"
const PT_VERTICAL_ALIGN = "-pt-valign"
const PT_FIT_CONTENT = "-pt-fit-content"
const PT_RESIZE_INSTANCES = "-pt-resize-instances"
const PT_OVERRIDE_SYMBOL = "-pt-override-symbol"
const PT_ATTR = "-pt-attr"

const SPACE_COLORS = ".--COLORS-"

const THIS_NAME = "_This"


// source: https://sketchplugins.com/d/956-set-fix-height-layer-property
const edgeFixdMap = {
    '-pt-pin-right': 1,
    '-pt-fix-size-width': 2,
    '-pt-pin-left': 4,
    '-pt-pin-bottom': 8,
    '-pt-fix-size-height': 16,
    '-pt-pin-top': 32,
}

const FIXED_SIZE_BEHAVIOUR_MAP = {
    'auto-width': 0,
    'auto-height': 1,
    'fixed-size': 2
}


const BLENDING_MODE_CSS_TO_SKETCH = {
    "normal": Style.BlendingMode.Normal,
    "darken": Style.BlendingMode.Darken,
    "multiply": Style.BlendingMode.Multiply,
    "color-burn": Style.BlendingMode.ColorBurn,
    "lighten": Style.BlendingMode.Lighten,
    "screen": Style.BlendingMode.Screen,
    "color-dodge": Style.BlendingMode.ColorDodge,
    "overlay": Style.BlendingMode.Overlay,
    //"darken": Style.BlendingMode.SoftLight, // Not supported in CSS
    //"darken": Style.BlendingMode.HardLight, // Not supported in CSS
    "difference": Style.BlendingMode.Difference,
    "exclusion": Style.BlendingMode.Exclusion,
    "hue": Style.BlendingMode.Hue,
    "saturation": Style.BlendingMode.Saturation,
    "color": Style.BlendingMode.Color,
    "luminosity": Style.BlendingMode.Luminosity,
}


const alignMap = {
    left: Text.Alignment.left,
    center: Text.Alignment.center,
    right: Text.Alignment.right,
    justify: Text.Alignment.justify
}
const alignMap2 = {
    [Text.Alignment.left]: "left",
    [Text.Alignment.center]: "center",
    [Text.Alignment.right]: "right",
    [Text.Alignment.justify]: "justify"
}
const vertAlignMap = {
    "top": Text.VerticalAlignment.top,
    "middle": Text.VerticalAlignment.center,
    "bottom": Text.VerticalAlignment.bottom
}
const vertAlignMap2 = {
    [Text.VerticalAlignment.top]: "top",
    [Text.VerticalAlignment.center]: "middle",
    [Text.VerticalAlignment.bottom]: "bottom",
}
const bordedLineEndMap = {
    "butt": Style.LineEnd.Butt,
    "round": Style.LineEnd.Round,
    "projecting": Style.LineEnd.Projecting
}

const bordedLineJoinMap = {
    "miter": Style.LineJoin.Miter,
    "round": Style.LineJoin.Round,
    "bevel": Style.LineJoin.Bevel
}

const smartLayoutMap = {
    "none": null,
    "left": SmartLayout.LeftToRight,
    "lefttoright": SmartLayout.LeftToRight,
    "center": SmartLayout.HorizontallyCenter,
    "horizontallycenter": SmartLayout.HorizontallyCenter,
    "right": SmartLayout.RightToLeft,
    "righttoleft": SmartLayout.RightToLeft,
    "toptobottom": SmartLayout.TopToBottom,
    "verticallycenter": SmartLayout.VerticallyCenter,
    "bottomtotop": SmartLayout.BottomToTop,
}

const bordedArrowheadMap = {
    "none": Style.Arrowhead.None,
    "openarrow": Style.Arrowhead.OpenArrow,
    "filledarrow": Style.Arrowhead.FilledArrow,
    "opencircle": Style.Arrowhead.OpenCircle,
    "filledcircle": Style.Arrowhead.FilledCircle,
    "opensquare": Style.Arrowhead.OpenSquare,
    "filledsquare": Style.Arrowhead.FilledSquare,
}


const weights = [
    { label: 'thin', sketch: 2, css: 100, title: "Thin" },
    { label: 'extra-light', sketch: 3, css: 200, title: "Extra Light" },
    { label: 'light', sketch: 4, css: 300, title: "Light" },
    { label: 'regular', sketch: 5, css: 400, title: "Regular" },
    { label: 'medium', sketch: 6, css: 500, title: "Medium" },
    { label: 'semi-bold', sketch: 8, css: 600, title: "Semi Bold" },
    { label: 'semibold', sketch: 8, css: 600, title: "Semi Bold#2" },
    { label: 'bold', sketch: 9, css: 700, title: "Bold" },
    { label: 'extra-bold', sketch: 10, css: 800, title: "Extra Bold" },
    { label: 'black', sketch: 12, css: 900, title: "Black" },
    { label: 'fa-solid', sketch: 14, css: "fa-solid", title: "Font Awesome Solid" },
]

// source: https://github.com/bahamas10/css-color-names/blob/master/css-color-names.json
const COLOR_NAMES = {
    "aliceblue": "#f0f8ff",
    "antiquewhite": "#faebd7",
    "aqua": "#00ffff",
    "aquamarine": "#7fffd4",
    "azure": "#f0ffff",
    "beige": "#f5f5dc",
    "bisque": "#ffe4c4",
    "black": "#000000",
    "blanchedalmond": "#ffebcd",
    "blue": "#0000ff",
    "blueviolet": "#8a2be2",
    "brown": "#a52a2a",
    "burlywood": "#deb887",
    "cadetblue": "#5f9ea0",
    "chartreuse": "#7fff00",
    "chocolate": "#d2691e",
    "coral": "#ff7f50",
    "cornflowerblue": "#6495ed",
    "cornsilk": "#fff8dc",
    "crimson": "#dc143c",
    "cyan": "#00ffff",
    "darkblue": "#00008b",
    "darkcyan": "#008b8b",
    "darkgoldenrod": "#b8860b",
    "darkgray": "#a9a9a9",
    "darkgreen": "#006400",
    "darkgrey": "#a9a9a9",
    "darkkhaki": "#bdb76b",
    "darkmagenta": "#8b008b",
    "darkolivegreen": "#556b2f",
    "darkorange": "#ff8c00",
    "darkorchid": "#9932cc",
    "darkred": "#8b0000",
    "darksalmon": "#e9967a",
    "darkseagreen": "#8fbc8f",
    "darkslateblue": "#483d8b",
    "darkslategray": "#2f4f4f",
    "darkslategrey": "#2f4f4f",
    "darkturquoise": "#00ced1",
    "darkviolet": "#9400d3",
    "deeppink": "#ff1493",
    "deepskyblue": "#00bfff",
    "dimgray": "#696969",
    "dimgrey": "#696969",
    "dodgerblue": "#1e90ff",
    "firebrick": "#b22222",
    "floralwhite": "#fffaf0",
    "forestgreen": "#228b22",
    "fuchsia": "#ff00ff",
    "gainsboro": "#dcdcdc",
    "ghostwhite": "#f8f8ff",
    "goldenrod": "#daa520",
    "gold": "#ffd700",
    "gray": "#808080",
    "green": "#008000",
    "greenyellow": "#adff2f",
    "grey": "#808080",
    "honeydew": "#f0fff0",
    "hotpink": "#ff69b4",
    "indianred": "#cd5c5c",
    "indigo": "#4b0082",
    "ivory": "#fffff0",
    "khaki": "#f0e68c",
    "lavenderblush": "#fff0f5",
    "lavender": "#e6e6fa",
    "lawngreen": "#7cfc00",
    "lemonchiffon": "#fffacd",
    "lightblue": "#add8e6",
    "lightcoral": "#f08080",
    "lightcyan": "#e0ffff",
    "lightgoldenrodyellow": "#fafad2",
    "lightgray": "#d3d3d3",
    "lightgreen": "#90ee90",
    "lightgrey": "#d3d3d3",
    "lightpink": "#ffb6c1",
    "lightsalmon": "#ffa07a",
    "lightseagreen": "#20b2aa",
    "lightskyblue": "#87cefa",
    "lightslategray": "#778899",
    "lightslategrey": "#778899",
    "lightsteelblue": "#b0c4de",
    "lightyellow": "#ffffe0",
    "lime": "#00ff00",
    "limegreen": "#32cd32",
    "linen": "#faf0e6",
    "magenta": "#ff00ff",
    "maroon": "#800000",
    "mediumaquamarine": "#66cdaa",
    "mediumblue": "#0000cd",
    "mediumorchid": "#ba55d3",
    "mediumpurple": "#9370db",
    "mediumseagreen": "#3cb371",
    "mediumslateblue": "#7b68ee",
    "mediumspringgreen": "#00fa9a",
    "mediumturquoise": "#48d1cc",
    "mediumvioletred": "#c71585",
    "midnightblue": "#191970",
    "mintcream": "#f5fffa",
    "mistyrose": "#ffe4e1",
    "moccasin": "#ffe4b5",
    "navajowhite": "#ffdead",
    "navy": "#000080",
    "oldlace": "#fdf5e6",
    "olive": "#808000",
    "olivedrab": "#6b8e23",
    "orange": "#ffa500",
    "orangered": "#ff4500",
    "orchid": "#da70d6",
    "palegoldenrod": "#eee8aa",
    "palegreen": "#98fb98",
    "paleturquoise": "#afeeee",
    "palevioletred": "#db7093",
    "papayawhip": "#ffefd5",
    "peachpuff": "#ffdab9",
    "peru": "#cd853f",
    "pink": "#ffc0cb",
    "plum": "#dda0dd",
    "powderblue": "#b0e0e6",
    "purple": "#800080",
    "rebeccapurple": "#663399",
    "red": "#ff0000",
    "rosybrown": "#bc8f8f",
    "royalblue": "#4169e1",
    "saddlebrown": "#8b4513",
    "salmon": "#fa8072",
    "sandybrown": "#f4a460",
    "seagreen": "#2e8b57",
    "seashell": "#fff5ee",
    "sienna": "#a0522d",
    "silver": "#c0c0c0",
    "skyblue": "#87ceeb",
    "slateblue": "#6a5acd",
    "slategray": "#708090",
    "slategrey": "#708090",
    "snow": "#fffafa",
    "springgreen": "#00ff7f",
    "steelblue": "#4682b4",
    "tan": "#d2b48c",
    "teal": "#008080",
    "thistle": "#d8bfd8",
    "tomato": "#ff6347",
    "turquoise": "#40e0d0",
    "violet": "#ee82ee",
    "wheat": "#f5deb3",
    "white": "#ffffff",
    "whitesmoke": "#f5f5f5",
    "yellow": "#ffff00",
    "yellowgreen": "#9acd32"
}

function degToRad (deg) {
    return deg * Math.PI / 180;
}